Create enos.py, Util file

amuraleedhar opened the pull request:
##### SUMMARY
Utility file for the modules for enos from Lenovo.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/enos.py

##### ANSIBLE VERSION
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible-2.5.0-py2.7.egg/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]


##### ADDITIONAL INFORMATION

This is an effort to create modules for eos based switches from Lenovo.

Commit:
Create enos.py, Util file

ganeshrn commented on lib/ansible/module_utils/enos.py:
```diff
@@ -0,0 +1,136 @@
+                      no_log=True),
+    'timeout': dict(type='int'),
+    'provider': dict(type='dict'),
+}
```
From 2.5 onwards provider arguments are added as suboption spec for all Network platform that uses persistent connection framework. Please refer [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ios.py#L45) for example implementation.

ganeshrn commented on lib/ansible/module_utils/enos.py:
```diff
@@ -0,0 +1,136 @@
+}
+
+
+def get_argspec():
```
Rename this to `get_provider_argspec` to be in sync with other platforms

ganeshrn commented on lib/ansible/module_utils/enos.py:
```diff
@@ -0,0 +1,136 @@
+
+
+def check_args(module, warnings):
+    provider = module.params['provider'] or {}
```
This check is not required as it is handled as part of suboption validation (after the provider change mentioned in above comment)

ganeshrn commented on lib/ansible/module_utils/enos.py:
```diff
@@ -0,0 +1,136 @@
+    try:
+        return _DEVICE_CONFIGS[cmd]
+    except KeyError:
+        rc, out, err = exec_command(module, cmd)
```
From 2.5 we are transitioning to use cliconf plugin and is currently work in progress. Please have a look at the ASA modules as they are the only ones (as of 2.4) converted to using cliconf.

ganeshrn replied:
With cliconf the `get_config()` implementation and other similar api implementation is moved to platform-specific cliconf plugins [here](https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/cliconf).

ganeshrn commented on lib/ansible/module_utils/enos.py:
```diff
@@ -0,0 +1,136 @@
+    commands = to_commands(module, to_list(commands))
+    for cmd in commands:
+        cmd = module.jsonify(cmd)
+        rc, out, err = exec_command(module, cmd)
```
Same as above.

ganeshrn commented on lib/ansible/module_utils/enos.py:
```diff
@@ -0,0 +1,136 @@
+    for command in to_list(commands):
+        if command == 'end':
+            continue
+        rc, out, err = exec_command(module, command)
```
Same as above.

Commit:
Editing based on comments of ganesh

Inspired from ASA modules

Commit:
Fixing Bugs on 2 validation Bugs

2017-10-10 07:48:17 ERROR: Found 2 pylint issue(s) which need to be resolved:
2017-10-10 07:48:17 ERROR: lib/ansible/module_utils/enos.py:74:0: dangerous-default-value Dangerous default value [] as argument (100%)
2017-10-10 07:48:17 ERROR: lib/ansible/module_utils/enos.py:125:21: undefined-variable Undefined variable 'config' (100%)

ganeshrn commented on lib/ansible/module_utils/enos.py:
```diff
@@ -0,0 +1,140 @@
+        return _DEVICE_CONFIGS[cmd]
+    except KeyError:
+        conn = get_connection(module)
+        out = conn.get(cmd)
```
Since enos  cliconf plugin has a specific `get_config()` api is it possible to move L75-83 to cliconf plugin and call `get_config()` api instead of `get()`
